lib/ref_arch_setup/install.rb

```diff
@@ -25,61 +31,307 @@ def initialize(target_master)
+    #
+    def download_pe_tarball(url, nodes)
+      puts "Attempting to download #{url} to #{nodes}"
+      puts
+
+      params = {}
+      params["url"] = url
+      params["destination"] = TMP_WORK_DIR
+
+      success = BoltHelper.run_task_with_bolt("ref_arch_setup::download_pe_tarball", params, nodes)
+      return success
+    end
+
+    # Downloads the PE tarball locally and moves it to the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The pe tarball URL
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def download_and_move_pe_tarball(url)
+      download_path = "#{TMP_WORK_DIR}/#{@pe_tarball_filename}"
+      success = download_pe_tarball(url, "localhost")
+      success = upload_pe_tarball(download_path) if success
       return success
     end
-    # rubocop:enable Metrics/PerceivedComplexity
+
+    # Handles the specified PE tarball URL by either downloading directly to the
+    # target master or downloading locally and copying to the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The PE tarball URL
+    #
+    # @return [string] The tarball path on the target master
+    def handle_tarball_url(url)
+      parse_url(url)
+      tarball_path_on_master = "#{TMP_WORK_DIR}/#{@pe_tarball_filename}"
+      remote_error = "Failed downloading #{url} to localhost and moving to #{@target_master}"
+
+      if target_master_is_localhost?
+        success = download_pe_tarball(url, "localhost")
+        raise "Failed downloading #{url} to localhost" unless success
+      else
+        # if downloading to the target master fails try to download locally and then upload
+        success = download_pe_tarball(url, @target_master)
+        puts "Unable to download the tarball directly to localhost" unless success
+        success = download_and_move_pe_tarball(url) unless success
+        raise remote_error unless success
+      end
+
+      return tarball_path_on_master
+    end
+
+    # Copies the PE tarball from the specified location to the temp working directory
+    # on the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path The pe tarball path
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def copy_pe_tarball_on_target_master(path)
+      command = "cp #{path} #{TMP_WORK_DIR}"
+      success = BoltHelper.run_cmd_with_bolt(command, @target_master)
+      return success
+    end
+
+    # Handles the specified tarball path when the target master is not localhost
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path The pe tarball path
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def handle_tarball_path_with_remote_target_master(path)
+      remote_flag = "#{@target_master}:"
+
+      if path.start_with?(remote_flag)
+        actual_path = path.sub!(remote_flag, "")
+        success = file_exist_on_target_master?(actual_path)
```

Comment:
we really shouldn't need two bolt calls to do this... But combining them would require processing output from the bolt run which we don't currently do.  So maybe add a TODO?  Or a ticket to combine these in the future.

Reply:
Created SLV-187: Combine file_exist_on_target_master? and copy_pe_tarball_on_target_master